---
title: "Advanced Bioinformatics 2023 assessment"
author: "m1903050"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown: General R/Rstudio assessment 

Task 3.1 
Using the sum() function and : operator, write an expression in the code snippet to evaluate the sum of all integers between 5 and 55. (4pt)
```{r}
sum(5:55)
```

Task 3.2
Write a function called sumfun with one input parameter, called n, that calculates the sum of all integers between 5 and n. Use the function to do the calculation for n = 10, n = 20, and n = 100 and present the results. (4pt)
```{r}
sumfun <- function(n)
  {sum(5:n)}

sumfun(10)
sumfun(20)
sumfun(100)
```

Task 3.3
The famous Fibonacci series is calculated as the sum of the two preceding members of the sequence, where the first two steps in the sequence are 1, 1. Write an R script using a for loop to calculate and print out the first 12 entries of the Fibonacci series. (4pt)
```{r}
n <- 12

fibonacci <- numeric(n) 
fibonacci[1] <- 1
fibonacci[2] <- 1

for (f in 3:n) 
  fibonacci[f] <- fibonacci[f-1]+fibonacci[f-2]

print(fibonacci)
```

Task 3.4
With the mtcars data set bundled with R, use ggplot to generate a box of miles per gallon (in the variable mpg) as a function of the number of gears (in the variable gear). Use the fill aesthetic to color bars by number of gears. (4pt)
```{r}
library(ggplot2)
ggplot(mtcars, aes(x= as.factor(gear),y=mpg))+
  geom_boxplot(aes(fill=as.factor(gear)))+
  labs(title="Miles per gallon as a function of the number of gears", x="Number of gears", y="Miles per gallon")
```

Task 3.5
Using the cars dataset and the function lm, fit a linear relationship between speed and breaking distance in the variable distance. What are the fitted slope and intercept of the line, and their standard errors? What are the units used for the variables in the dataset? (4pt)
```{r}
# The head command was used to determine and explore the data set with its corresponding variables.
head(cars, 5)

# The linear_model variable is created from the lm function of the distance proportional to the speed.
linear_model <- lm(formula = dist ~ speed, data=cars)

# This line provides the summary of the coefficients in the linear_model variable. 
summary(linear_model)$coefficients

# The Y intercept of the line has a value of -17.579095 whereas the fitted slope of the breaking distance has a value of 3.932409.

# The standard error for the Y intercept is 6.7584402 whereas the fitted slope of the breaking distance has a standard error value of 0.4155128.

# The units for the speed variable is in miles per hour (mph) whereas the units used for the breaking distant variable is in feet (ft).
```

Task 3.6
Use ggplot to plot the data points from Task 6 and the linear fit. (4pt)
```{r}
ggplot(cars, aes(x = speed, y=dist))+
  geom_point()+
  geom_smooth(method = "lm")+
  labs(title="Linear Relationship between Speed and Breaking Distance",
       x="Speed in miles per hour (mph)",
       y="Breaking Distance in feet (ft)")
```

Task 3.7
Again using the cars dataset, now use linear regression (lm) to estimate the average reaction time for the driver to start breaking (in seconds). To simplify matters you may assume that once breaking commences, breaking distance is proportional to the square of the speed. Explain the steps in your analysis. Do you get reasonable results? Finally, use ggplot to plot the data points and the fitted relationship. (9pt)
```{r}
# 1 feet = 0.000189394 mile; the breaking distance is multiplied by 0.000189394 to convert the distant variables to miles.
breakdis_mil <- cars$dist*0.000189394

# As the breaking distance is proportional to the square of the speed, the variable speedsqr was made. 
speedsqr <- cars$speed^2

# The 'lm' function is then used to find the coefficients of the between the breaking distance in miles and the square of the speeds.
Reaction_time_linear_model <- lm(formula = breakdis_mil ~ speedsqr)
Reaction_time_linear_model

# The fitted slope coefficient was then extracted and placed in the variable 'fitted_co' as it is represents the expected breaking distance per mile added. 
fitted_co <- coef(Reaction_time_linear_model)[2]
fitted_co

# The mean values of the speed  are placed in a set variable 'speed_mean'.
speed_mean <- mean(cars$speed)
speed_mean

# The reaction distance is assumed to be approximately half that of the breaking distance; the stopping distance is the sum of the breaking distance and the reaction distance. Therefore to calculate the average reaction distance we would divide the fitted_co variable by 2 and then multiply the value with the mean speed find the average reaction distance. This formula was then inputted into the 'av_react_dist' variable.
av_react_dist <- (fitted_co/2)*speed_mean
av_react_dist

# The average reaction time in hours is calculated by dividing the average breaking distance  with the 'av_react_dist' variable. 
rtime_hrs <- av_react_dist/speed_mean
rtime_hrs

# To convert 1 hour to seconds, we multiply 60 by 60 to get 3600 seconds in the variable 'hr_to_sec'.
hr_to_sec <- 60*60
hr_to_sec

# The reaction time was then converted to seconds by dividing it with the 'hr_to_sec' variable, this gives an average reaction time of 0.04396661 seconds.
rtime_sec <- rtime_hrs*hr_to_sec
rtime_sec

# The given reaction time in seconds appears to be lower than the expected average value of 1.5 seconds. Therefore the proposed calculation does not appear to be accurate as there are a number of factors required to calculate the reaction time of a moving vehicle. This includes environmental factors; human factors such as sex, age, and intoxications; type of    

# ggplot is used to plot the data using the custom variables described previously. 
ggplot(cars, aes(x = speedsqr, y=breakdis_mil))+
  geom_point()+
  geom_smooth(method = "lm")+
  labs(title = "Reaction time Linear Model",
       x = expression("speed in miles per hour squared"~mph^2),
       y = "Breaking distance in miles")
```

## R Markdown: RNA-seq assessment 

Task 3.8
Read in count data and sample description. (1pts)
```{r}
# Read input file with exercise count data
exercise_counts <- read.csv(file="LMS_RNAseq_short-master-2023-final/course/exercises/data/exercise1_counts.csv", row.names = 1)

# explore exercise count data
head(exercise_counts)

# Read input file with sample description
sample_des <- read.table("LMS_RNAseq_short-master-2023-final/course/exercises/data/exercise1_sample_description.info", sep= "\t", header = TRUE)

# Explore sample description data
head(sample_des)
```

Task 3.9
Create col_data and check dimensions.  (1 pts)
```{r}
# Prepare data for DESeq
col_data <- data.frame(sample = sample_des$sample,
                       condition = sample_des$condition,
                       batch = sample_des$batch)

# Store as factors
col_data$sample <- as.factor(col_data$sample)
col_data$condition <- as.factor(col_data$condition)
col_data$batch <- as.factor(col_data$batch)

# Check dimensions
all(colnames(exercise_counts) == sample_des$sample)
all(colnames(exercise_counts) == sample_des$condition)
all(colnames(exercise_counts) == sample_des$batch)
```

Task 3.10
Construct DESeqDataSet object using count data and sample description.   (1 pts)
```{r}
# Load DESeq2 library
library(DESeq2)

dds <- DESeqDataSetFromMatrix(countData = exercise_counts, 
                              colData = col_data, 
                              design = ~condition)

# DESeq normalization
dds <- DESeq(dds)
```

Task 3.11
Perform rlog and VST transformation on the data. (1 pts) 
```{r}
# rlog transformation
rld <- rlog(dds)

# Get rld in count format
rld_counts <- assay(rld)

# VST transformation
vsd <- varianceStabilizingTransformation(dds)

# Get rld in count format
vsd_counts <- assay(vsd)
```

Task 3.12
Draw a heatmap of count matrix based on the top 40 highly expressed genes using rlog and VST data. (1 pts)
```{r}
# Load pheatmap library
library("pheatmap")

# Get dds normalized counts
dds_counts <- counts(dds, normalized = TRUE)

# Get normalized counts - 40 higher values
select <- order(rowMeans(dds_counts), decreasing = TRUE)[1:40]

# Heatmap of the rlog transformed data
pheatmap(assay(rld)[select, ])

# Heatmap of the vst transformed data
pheatmap(assay(vsd)[select, ])
```

Task 3.13
Generate a SDM to see the clustering of count data. (1 pts)
```{r}
# Generate a sample distance matrix (SDM) from rlog transformed data
sample_dist <- dist(t(assay(rld)))
class(sample_dist)

# Get SDM in matrix form
sdm <- as.matrix(sample_dist)
class(sdm)

# Load library
library("RColorBrewer")

# Add row names for clear plot
rownames(sdm) <- rld$Group
colnames(sdm) <- NULL

# Add colors
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

# Plot heatmap
pheatmap(sdm,
         clustering_distance_rows = sample_dist,
         clustering_distance_cols = sample_dist,
         col = colors)
```

Task 3.14
Perform the Principal Component Analysis using rlog method and find out the % significance values of first two principal components. (1 pts)
```{r}
# PCA plot on our rld transformed data
plotPCA(rld, intgroup = "condition")
```

Task 3.15
Repeat the PCA, this time using VST method and compare the plots with the ones obtained using rlog method. (1 pts)
```{r}
# PCA plot on our vst transformed data
plotPCA(vsd, intgroup = "condition")
```

## R Markdown: ChIP-seq assessment

Task 3.16
Read in the two Myc Mel peakset replicates and create the common peakset as we did for our previous exercise. (1 pts) 
```{r}

peakset1 <- ChIPQC:::GetGRanges("LMS_ChIPseq_short-master-2023-final/course/data/MacsPeaks/mycmelrep1_peaks.xls", sep="\t", simple=F)
peakset2 <- ChIPQC:::GetGRanges("LMS_ChIPseq_short-master-2023-final/course/data/MacsPeaks/mycmelrep2_peaks.xls", sep="\t", simple=F)

allpeaks <- c(peakset1,peakset2)
allpeaksreduced <- reduce(allpeaks)
allpeaksreduced

common_peakset <- allpeaksreduced[allpeaksreduced %over% peakset1
                                  & allpeaksreduced %over% peakset2]

length(common_peakset)

common_peakset[1:500,]
```

Task 3.17
Now we can rank them by their fold enrichment, select the top 500 peaks and resize these peaks to 200bp around centre. (1 pts)
```{r}
allpeaks <- c(peakset1,peakset2)
allpeaks


allpeaks_ranked <- allpeaks[rev(order(allpeaks$fold_enrichment))]
allpeaks_ranked


commonpeaksreduced <- reduce(allpeaks_ranked)
commonpeaksreduced


top_500 <- commonpeaksreduced[1:500]
top_500


commonpeaks <- resize(top_500,200,fix="center") 
commonpeaks
```

Task 3.18
Extract the sequences underneath the file and write them to FASTA file in you working directory. Inspect the file in notepad. (1 pts)
```{r}
library(BSgenome)
library(BSgenome.Mmusculus.UCSC.mm9)

genome <- BSgenome.Mmusculus.UCSC.mm9

seqlevelsStyle(commonpeaks) <- "UCSC"


commonpeaksequences <- getSeq(genome,GRanges(commonpeaks))

names(commonpeaksequences) <-  paste0("peak_",seqnames(commonpeaks),"_",
                                      start(commonpeaks),
                                      "-",
                                      end(commonpeaks))

commonpeaksequences[1:2]

writeXStringSet(commonpeaksequences,file="consensusPeaks.fa")

```

Task 3.19
Upload the sequences to Meme-ChIP and report the results when complete. (1 pts)

================================================================================

Starting MEME-ChIP

meme-chip -oc . -time 240 -ccut 100 -dna -order 2 -minw 6 -maxw 15 -db db/motif_databases/MOUSE/uniprobe_mouse.meme -db db/motif_databases/JASPAR/JASPAR2022_CORE_vertebrates_non-redundant_v2.meme -db db/motif_databases/EUKARYOTE/jolma2013.meme -meme-mod zoops -meme-nmotifs 3 -meme-searchsize 100000 -streme-pvt 0.05 -streme-totallength 4000000 -centrimo-score 5.0 -centrimo-ethresh 10.0 consensusPeaks.fa

MEME-ChIP is starting subprocess getsize

getsize ./consensusPeaks.fa 1> $metrics

MEME-ChIP subprocess getsize ran successfully in 0.0 seconds

MEME-ChIP is starting subprocess fasta-most

fasta-most -min 50 < ./consensusPeaks.fa 1> $metrics

MEME-ChIP subprocess fasta-most ran successfully in 0.0 seconds

MEME-ChIP is starting subprocess fasta-center

fasta-center -dna -len 100 < ./consensusPeaks.fa 1> ./seqs-centered

MEME-ChIP subprocess fasta-center ran successfully in 0.0 seconds

MEME-ChIP is starting subprocess fasta-get-markov

fasta-get-markov -nostatus -nosummary -dna -m 2 ./consensusPeaks.fa ./background

MEME-ChIP subprocess fasta-get-markov ran successfully in 0.0 seconds

MEME-ChIP is starting subprocess meme

meme ./seqs-centered -oc meme_out -mod zoops -nmotifs 3 -minw 6 -maxw 15 -bfile ./background -dna -searchsize 100000 -time 2919 -revcomp 
-nostatus

MEME-ChIP subprocess meme ran successfully in 317.2 seconds

MEME-ChIP is starting subprocess streme

streme --verbosity 1 --oc streme_out -dna --p ./seqs-centered --minw 6 --maxw 15 --time 6063 --order 2 --thresh 0.05 --totallength 4000000

MEME-ChIP subprocess streme ran successfully in 4.7 seconds

MEME-ChIP is starting subprocess centrimo

centrimo -seqlen 200 -verbosity 1 -oc centrimo_out -bfile ./background -score 5.0 -ethresh 10.0 ./consensusPeaks.fa meme_out/meme.xml 
streme_out/streme.xml db/motif_databases/MOUSE/uniprobe_mouse.meme db/motif_databases/JASPAR/JASPAR2022_CORE_vertebrates_non-redundant_v2.meme db/motif_databases/EUKARYOTE/jolma2013.meme
MEME-ChIP subprocess centrimo ran successfully in 8.0 seconds

MEME-ChIP is starting subprocess tomtom

tomtom -verbosity 1 -oc meme_tomtom_out -min-overlap 5 -dist pearson -evalue -thresh 1 -no-ssc meme_out/meme.xml db/motif_databases/MOUSE/uniprobe_mouse.meme db/motif_databases/JASPAR/JASPAR2022_CORE_vertebrates_non-redundant_v2.meme db/motif_databases/EUKARYOTE/jolma2013.meme

MEME-ChIP subprocess tomtom ran successfully in 9.2 seconds

MEME-ChIP is starting subprocess tomtom

tomtom -verbosity 1 -oc streme_tomtom_out -min-overlap 5 -dist pearson -evalue -thresh 1 -no-ssc streme_out/streme.xml db/motif_databases/MOUSE/uniprobe_mouse.meme db/motif_databases/JASPAR/JASPAR2022_CORE_vertebrates_non-redundant_v2.meme db/motif_databases/EUKARYOTE/jolma2013.meme

MEME-ChIP subprocess tomtom ran successfully in 5.7 seconds

MEME-ChIP is starting subprocess tomtom

tomtom -verbosity 1 -text -thresh 0.1 ./combined.meme ./combined.meme 1> ./motif_alignment.txt

MEME-ChIP subprocess tomtom ran successfully in 0.0 seconds

MEME-ChIP is starting subprocess spamo

spamo -verbosity 1 -oc spamo_out_1 -bgfile ./background -keepprimary -primary TGAGCCACCATGTGG ./consensusPeaks.fa meme_out/meme.xml meme_out/meme.xml streme_out/streme.xml

MEME-ChIP subprocess spamo ran successfully in 0.1 seconds

MEME-ChIP is starting subprocess spamo

spamo -verbosity 1 -oc spamo_out_2 -bgfile ./background -keepprimary -primary GATCTCATTACAGAT ./consensusPeaks.fa meme_out/meme.xml meme_out/meme.xml streme_out/streme.xml

MEME-ChIP subprocess spamo ran successfully in 0.1 seconds

MEME-ChIP is starting subprocess fimo

fimo --parse-genomic-coord --verbosity 1 --oc fimo_out_1 --bgfile ./background --motif TGAGCCACCATGTGG meme_out/meme.xml ./consensusPeaks.fa

MEME-ChIP subprocess fimo ran successfully in 0.1 seconds

MEME-ChIP is starting subprocess fimo

fimo --parse-genomic-coord --verbosity 1 --oc fimo_out_2 --bgfile ./background --motif GATCTCATTACAGAT meme_out/meme.xml ./consensusPeaks.fa

MEME-ChIP subprocess fimo ran successfully in 0.1 seconds

MEME-ChIP is starting subprocess meme-chip_html_to_tsv

meme-chip_html_to_tsv ./meme-chip.html ./summary.tsv "meme-chip -oc . -time 240 -ccut 100 -dna -order 2 -minw 6 -maxw 15 -db db/motif_databases/MOUSE/uniprobe_mouse.meme -db db/motif_databases/JASPAR/JASPAR2022_CORE_vertebrates_non-redundant_v2.meme -db db/motif_databases/EUKARYOTE/jolma2013.meme -meme-mod zoops -meme-nmotifs 3 -meme-searchsize 100000 -streme-pvt 0.05 -streme-totallength 4000000 -centrimo-score 5.0 -centrimo-ethresh 10.0 consensusPeaks.fa" 5.5.2 "Sun Jan 29 10\:33\:12 2023 -0800"

MEME-ChIP subprocess meme-chip_html_to_tsv ran successfully in 0.1 seconds

MEME-ChIP ran successfully in 345.67 seconds

Starting Tar
tar -czf appMEMECHIP_5.5.21681774911062596627524.tar.gz -C /data/apache-tomcat/instance/meme/webapps/opal-jobs 
appMEMECHIP_5.5.21681774911062596627524/fimo_out_2/fimo.html appMEMECHIP_5.5.21681774911062596627524/meme_out/logo_rc3.png 
appMEMECHIP_5.5.21681774911062596627524/spamo_out_2/spamo.html appMEMECHIP_5.5.21681774911062596627524/streme_out/streme.html 
appMEMECHIP_5.5.21681774911062596627524/streme_out/streme.txt appMEMECHIP_5.5.21681774911062596627524/summary.tsv 
appMEMECHIP_5.5.21681774911062596627524/centrimo_out/centrimo.html appMEMECHIP_5.5.21681774911062596627524/meme_out/logo1.eps 
appMEMECHIP_5.5.21681774911062596627524/meme_out/logo_rc2.eps appMEMECHIP_5.5.21681774911062596627524/combined.meme 
appMEMECHIP_5.5.21681774911062596627524/fimo_out_1/fimo.gff appMEMECHIP_5.5.21681774911062596627524/messages.txt 
appMEMECHIP_5.5.21681774911062596627524/spamo_out_2/spamo.tsv appMEMECHIP_5.5.21681774911062596627524/meme_out/logo_rc3.eps 
appMEMECHIP_5.5.21681774911062596627524/spamo2_msgs.txt appMEMECHIP_5.5.21681774911062596627524/meme_out/logo3.png 
appMEMECHIP_5.5.21681774911062596627524/meme_out/meme.html appMEMECHIP_5.5.21681774911062596627524/meme_out/logo2.eps 
appMEMECHIP_5.5.21681774911062596627524/meme_out/logo2.png appMEMECHIP_5.5.21681774911062596627524/streme_out/streme.xml 
appMEMECHIP_5.5.21681774911062596627524/fimo_out_1/fimo.tsv appMEMECHIP_5.5.21681774911062596627524/spamo_out_1/spamo.html 
appMEMECHIP_5.5.21681774911062596627524/seqs-centered appMEMECHIP_5.5.21681774911062596627524/progress_log.txt 
appMEMECHIP_5.5.21681774911062596627524/meme_out/logo3.eps appMEMECHIP_5.5.21681774911062596627524/fimo_out_2/fimo.tsv 
appMEMECHIP_5.5.21681774911062596627524/fimo_out_2/fimo.xml appMEMECHIP_5.5.21681774911062596627524/fimo_out_1/fimo.xml 
appMEMECHIP_5.5.21681774911062596627524/meme_out/logo1.png appMEMECHIP_5.5.21681774911062596627524/streme_out/sequences.tsv 
appMEMECHIP_5.5.21681774911062596627524/meme_out/meme.txt appMEMECHIP_5.5.21681774911062596627524/centrimo_out/site_counts.txt 
appMEMECHIP_5.5.21681774911062596627524/background appMEMECHIP_5.5.21681774911062596627524/meme_out/logo_rc2.png 
appMEMECHIP_5.5.21681774911062596627524/fimo_out_1/cisml.xml appMEMECHIP_5.5.21681774911062596627524/meme_tomtom_out/tomtom.xml 
appMEMECHIP_5.5.21681774911062596627524/meme_out/logo_rc1.png appMEMECHIP_5.5.21681774911062596627524/meme_out/meme.xml 
appMEMECHIP_5.5.21681774911062596627524/streme_tomtom_out/tomtom.xml appMEMECHIP_5.5.21681774911062596627524/motif_alignment.txt 
appMEMECHIP_5.5.21681774911062596627524/fimo_out_1/fimo.html appMEMECHIP_5.5.21681774911062596627524/meme_out/logo_rc1.eps 
appMEMECHIP_5.5.21681774911062596627524/spamo1_msgs.txt appMEMECHIP_5.5.21681774911062596627524/meme-chip.html 
appMEMECHIP_5.5.21681774911062596627524/spamo_out_1/spamo.tsv appMEMECHIP_5.5.21681774911062596627524/meme_tomtom_out/tomtom.html 
appMEMECHIP_5.5.21681774911062596627524/centrimo_out/centrimo.tsv appMEMECHIP_5.5.21681774911062596627524/fimo_out_2/fimo.gff 
appMEMECHIP_5.5.21681774911062596627524/streme_tomtom_out/tomtom.tsv appMEMECHIP_5.5.21681774911062596627524/fimo_out_2/cisml.xml 
appMEMECHIP_5.5.21681774911062596627524/meme_tomtom_out/tomtom.tsv appMEMECHIP_5.5.21681774911062596627524/streme_tomtom_out/tomtom.html 
appMEMECHIP_5.5.21681774911062596627524/consensusPeaks.fa

Tar ran successfully in 0.13 seconds

Done
================================================================================